### PR TITLE
support all conan build policies

### PR DESCRIPTION
**Pull request by boussaffawalid**
Changelog: (Feature): support all conan build policy from https://docs.conan.io/en/latest/mastering/policies.html

I ussue have been raised here https://github.com/conan-io/conan-package-tools/issues/394

- [ x] I've read the [Contributing guide](https://github.com/conan-io/conan-package-tools/blob/develop/.github/CONTRIBUTING.md).
- [ x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008) style guides for Python code.

### CODE REVIEW

**Comment by uilianries** [comment]
@boussaffawalid Thanks for your contribution. It would be great if you could add a test using multi build policies together.

**Comment by boussaffawalid** [comment]
@uilianries I've added couple of unittest. I got a failure however
```
Python 2 deprecation notice has been bypassed by envvar
'USE_UNSUPPORTED_CONAN_WITH_PYTHON_2'
ERROR: Specify the 'user' and the 'channel' or neither of them
```

I may need your help here, tests are passing on my dev machine. (I use pyton3)

File: cpt/test/integration/update_some_deps_test.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    def build(self):
+        self.output.warn("BUILDING")
+"""
+
+    def setUp(self):
+        self._ci_manager = MockCIManager()
+        self._server = TestServer(users={"user": "password"},
+                                  write_permissions=[("bar/0.1.0@foo/stable", "user"),
+                                                     ("foo/1.0.0@bar/testing", "user"),
+                                                     ("qux/1.0.0", "user")])
```
Comment (uilianries) [inline cpt/test/integration/update_some_deps_test.py:69]:
```suggestion
                                                     ("qux/1.0.0@qux/stable", "user")])
```

File: cpt/test/integration/update_some_deps_test.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        self._client = TestClient(servers={"default": self._server},
+                                  users={"default": [("user", "password")]})
+        self._client.save({"conanfile_bar.py": self.conanfile_bar})
+        self._client.run("export conanfile_bar.py foo/stable")
+        self._client.save({"conanfile_foo.py": self.conanfile_foo})
+        self._client.run("export conanfile_foo.py bar/testing")
+        self._client.save({"conanfile_foo3.py": self.conanfile_foo_3})
+        self._client.run("export conanfile_foo3.py")
```
Comment (uilianries) [inline cpt/test/integration/update_some_deps_test.py:77]:
```suggestion
        self._client.run("export conanfile_foo3.py qux/stable")
```

File: cpt/test/integration/update_some_deps_test.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                                                      build_policy=["foobar", "bar", "foo", "qux"],
+                                                      exclude_vcvars_precommand=True,
+                                                      ci_manager=self._ci_manager)
+            mulitpackager.add({}, {})
+            mulitpackager.run()
+
+            self.assertIn("Uploading packages for 'foobar/2.0@user/testing'", self._client.out)
+            self.assertIn("Uploading packages for 'bar/0.1.0@foo/stable'", self._client.out)
+            self.assertIn("Uploading packages for 'foo/1.0.0@bar/testing'", self._client.out)
+            self.assertIn("Uploading packages for 'qux/1.0.0@'", self._client.out)
```
Comment (uilianries) [inline cpt/test/integration/update_some_deps_test.py:98]:
```suggestion
            self.assertIn("Uploading packages for 'qux/1.0.0@qux/stable'", self._client.out)
```

File: cpt/test/integration/update_some_deps_test.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                                                      channel="testing",
+                                                      build_policy="foobar",
+                                                      exclude_vcvars_precommand=True,
+                                                      ci_manager=self._ci_manager,
+                                                      conanfile="conanfile.py")
+            mulitpackager.add({}, {})
+            mulitpackager.run()
+            self.assertRegex(str(self._client.out), r'bar/0.1.0@foo/stable:.* - Cache')
+            self.assertRegex(str(self._client.out), r'foo/1.0.0@bar/testing:.* - Cache')
+            self.assertRegex(str(self._client.out), r'qux/1.0.0:.* - Cache')
```
Comment (uilianries) [inline cpt/test/integration/update_some_deps_test.py:111]:
```suggestion
            self.assertRegex(str(self._client.out), r'qux/1.0.0@qux/stable:.* - Cache')
```

File: cpt/test/integration/update_some_deps_test.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+            self.assertRegex(str(self._client.out), r'foobar/2.0@user/testing:.* - Build')
+
+            self.assertIn("Uploading packages for 'foobar/2.0@user/testing'", self._client.out)
+            self.assertNotIn("Uploading packages for 'bar/0.1.0@foo/stable'", self._client.out)
+            self.assertNotIn("Uploading packages for 'foo/1.0.0@bar/testing'", self._client.out)
+            self.assertNotIn("Uploading packages for 'qux/1.0.0@'", self._client.out)
```
Comment (uilianries) [inline cpt/test/integration/update_some_deps_test.py:118]:
```suggestion
            self.assertNotIn("Uploading packages for 'qux/1.0.0@qux/stable'", self._client.out)
```

**Comment by uilianries** [comment]
assertRegex is not available in python 2.7. YOu can take a look here: https://docs.python.org/2.7/library/unittest.html

**Comment by uilianries** [comment]
Thanks a log @boussaffawalid !

**Comment by kaihowl** [comment]
Thank you for the addition. This is really helpful. Is there anything blocking a new release of cpt that would include this change?

**Comment by uilianries** [comment]
It's released following the Conan release. It should be done during this week

**Comment by Croydon** [comment]
By the way for people programmatically using/extending CPT this is a potential break as the return value of the build_policy is now a list

For Bincrafters this at least broke a test https://travis-ci.com/github/bincrafters/bincrafters-package-tools/jobs/347580990#L442

**Comment by uilianries** [comment]
@Croydon thanks for the information! CPT is under 0.x version, which means, we can break, but we should avoid of course. As Conan supports more than one build policy (cascade for example) I think that's a valid change. Updating Bincrafters Package Tools will be the best strategy I think

**Comment by uilianries** [comment]
@Croydon I see you already fixed and released a new version for Bincrafters package tools! Vielen Dank :heart:

**Pull request merged**